Deprecation of VFS Class Loader items, new Context class loader management

dlmarion opened the pull request.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/spi/common/ContextClassLoaderFactory.java:
```diff
@@ -0,0 +1,60 @@
+
+  static class ClassLoaderFactoryConfiguration {
+
+    public Iterator<Entry<String,String>> get() {
```
Other SPI interfaces use ServiceEnvironment to expose Accumulo configuration to plugins.

https://github.com/apache/accumulo/blob/80ee9ca8092dc4bf07aa1a046d6e668f0e16300e/core/src/main/java/org/apache/accumulo/core/spi/common/ServiceEnvironment.java#L35

dlmarion commented on core/src/main/java/org/apache/accumulo/core/classloader/ClassLoaderUtil.java:
```diff
@@ -0,0 +1,32 @@
+
+public class ClassLoaderUtil {
+
+  public static synchronized <U> Class<? extends U> loadClass(String contextName, String className,
```
Now that you bring it up, I'm not sure that this needs to be synchronized.

keith-turner replied:
Looking into this overall issue a bit further in the existing code, iterators call AccumuloVFSClassLoader.loadClass() which is static and synchronized.  Also the iterator loading code has a class cache, maybe this was created to avoid the global synchronization.  Also I did not find an issue, will open one after looking into it a bit more.

keith-turner replied:
The cache is only narrowly used, so the synchronization is probably still problematic I'll open an issue.

keith-turner replied:
Created #1730

ctubbsii commented on core/src/main/java/org/apache/accumulo/core/spi/common/ContextClassLoaderFactory.java:
```diff
@@ -0,0 +1,60 @@
+ */
+public interface ContextClassLoaderFactory {
+
+  static class ClassLoaderFactoryConfiguration {
```
They shouldn't inherit from the outer class, but it's also package-private, so it shouldn't matter.

keith-turner replied:
I think all members of interfaces are implicitly public.

ctubbsii replied:
> I think all members of interfaces are implicitly public.

Interesting. I hadn't considered if that might be true for classes defined in an interface. I know it applies to methods and fields. A quick Google search didn't help answer this for me, so I'll have to experiment to see if that holds for static inner classes.

dlmarion commented:
@keith-turner - I addressed some of the comments you made, others were OBE based on changes I made after the fact. Things are working now and this is out of draft state if you are interested in making another pass.

keith-turner commented on core/src/main/java/org/apache/accumulo/core/spi/common/ContextClassLoaderFactory.java:
```diff
@@ -0,0 +1,55 @@
+ */
+public interface ContextClassLoaderFactory {
+
+  /**
+   * Initialize the ClassLoaderFactory. Implementations may need a reference to the configuration so
+   * that it can clean up contexts that are no longer being used.
+   *
+   * @param contextProperties
+   *          Accumulo configuration properties
+   * @throws Exception
+   *           if error initializing ClassLoaderFactory
+   */
+  void initialize(Supplier<Map<String,String>> contextProperties) throws Exception;
```
The following are the rational behind the suggestion below.
 * Using a parameters interface makes evolving the SPI over time much simpler.  Its easy to add additional parameters in the future w/o breaking existing code that was written against the SPI.
 * Using the [Configuration](https://github.com/apache/accumulo/blob/353c611891a8b88b307dcba5e884467d4191b1f2/core/src/main/java/org/apache/accumulo/core/spi/common/ServiceEnvironment.java#L40) interface used by other SPI interface is consistent and offers more options.  For example this interface offers an isSet() method that allows checking if a user set a property.  There is no way to do this by analyzing the values because some props have default values.

```suggestion
  public interface InitParameters {
      /**
   * @return A view of Accumulo's system level configuration. This is backed by system level config
   *         in zookeeper, which falls back to site configuration, which falls back to the default
   *         configuration.
   */
    Configuration getConfiguration();
  }

  /**
   * Initialize the ClassLoaderFactory. Implementations may need a reference to the configuration so
   * that it can clean up contexts that are no longer being used.
   *
   * @throws Exception
   *           if error initializing ClassLoaderFactory
   */
  void initialize(InitParameters initParam) throws Exception;
```

dlmarion replied:
@keith-turner It looks like ServiceEnvironmentImpl and ConfigurationImpl are the classes used to create the Configuration used in this manner. Both of these objects are in server-base, which depends on accumulo-core. Suggestions on how to move forward? I think ContextClassLoaderFactory should remain in core as that is where the SPI classes are. It looks like I might be able to move ServiceEnvironmentImpl up to accumulo-core.

dlmarion replied:
I can't move ServiceEnvironmentImpl up to accumulo-core due to its depedency on ServerContext.

keith-turner replied:
I think what you did with movin ConfigurationImpl and using the ConfigurationImpl is good.  Would still recommend the parameters interface as it gives more options for mutating the SPI in the future w/o causing pain to anyone who has taken the time to implement the interface.

ctubbsii replied:
> Would still recommend the parameters interface as it gives more options for mutating the SPI in the future w/o causing pain to anyone who has taken the time to implement the interface.

I think I would disagree with the need to create an evolve-able parameter type here. I don't think it's a good idea to support a dedicated parameter type because that implies we're constructing a framework that can evolve, on which complex subclass implementations can support being initialized based on internal Accumulo objects. That's antithetical to what we're trying to do here.

This ClassloaderFactory is not your typical SPI (like sampling or iterators or balancers), where behavior is coupled to Accumulo functionality. Rather, classloading is something that should be done largely independently of Accumulo... as it is something on which other parts of Accumulo depend. Its initialization API should not itself depend on Accumulo objects that would be passed in using these parameter types, because that could create cyclical dependencies (Accumulo depends on classloader, which depends on Accumulo internal objects passed as parameters, which depends on classloader...).

The only thing this interface needs from Accumulo is initial configuration, which just happens to be carried in Accumulo's own configuration files. We should merely pass that information along to this. We should never pass along any other internal state objects as parameters. Besides, the classloader factory should be instantiated *early*, before any meaningful state is generated in Accumulo. So, it should never have the opportunity to initialize based on internal Accumulo state anyway... just initial configuration files.

So, I think it's a conscious decision to make the initialization parameter only a configuration mapping. It will ensure we keep the classloading stuff as decoupled as possible from Accumulo's internal workings, and merely responsible for providing classloaders that the rest of Accumulo can reliably depend on, based on the implementer's own independent criteria.

ctubbsii replied:
Actually, because classloading can get messy, I've debated arguing in favor of removing this initialize method entirely, and just expecting implementations rely on their own way of getting configuration (their own config files, or environment parsing), so it's even more decoupled from Accumulo. But... that would make our own provided implementations that are currently built-in (for preserving existing behavior) more difficult, because they wouldn't be able to read `general.classpaths`, etc.

keith-turner replied:
The concept that we know everything we may ever want to pass into a method is not something that has worked out too well in the past for the Accumulo API and its riddled with many duplicative overloaded methods as a result.   The parameter interface is an attempt to avoid repeating a mistake that has been made many times in the past.   

I don't think there is a 0% chance that another parameter will ever be needed.  While valid use case likely exists, we may or may not encounter them.  I have no idea what the chances are, so I am ok w/ someone making the judgement call that its not worth worrying about.  Personally I would worry about it simply based on past experience w/ API evolution.

ctubbsii replied:
It wasn't so much that I think "we know everything we may ever want to pass", so much as it was "I want to restrict what we can pass, to avoid scope creep". I don't feel too strongly about this and am okay with Keith's API strategy here.

ctubbsii commented on start/src/main/java/org/apache/accumulo/start/classloader/vfs/AccumuloVFSClassLoader.java:
```diff
@@ -358,48 +360,24 @@ public static void printClassPath(Printer out, boolean debug) {
           if (debug) {
-            out.print("Level " + classLoaderDescription + " " + classLoader.getClass().getName()
-                + " configuration not inspectable.\n");
+            out.print(classLoaderDescription + ": configuration not inspectable.\n");
```
I think there might be an IT that checks the printed strings... not sure. It's fine if it changes, but the IT might need to be updated.

ctubbsii reviewed (changes requested):
Thanks @dlmarion . I finally got a chance to look this over. Sorry it took so long to respond to the review request.

dlmarion commented on core/src/main/java/org/apache/accumulo/core/classloader/ClassLoaderUtil.java:
```diff
@@ -0,0 +1,34 @@
+
+  public static <U> Class<? extends U> loadClass(String contextName, String className,
+      Class<U> extension) throws ClassNotFoundException {
+    if (contextName != null && !contextName.equals(""))
```
Changed.

dlmarion commented on core/src/main/java/org/apache/accumulo/core/classloader/ClassLoaderUtil.java:
```diff
@@ -0,0 +1,34 @@
+
+public class ClassLoaderUtil {
+
+  public static <U> Class<? extends U> loadClass(String contextName, String className,
+      Class<U> extension) throws ClassNotFoundException {
+    if (contextName != null && !contextName.equals(""))
+      return ContextClassLoaders.getClassLoader(contextName).loadClass(className)
+          .asSubclass(extension);
+    else
+      return AccumuloVFSClassLoader.loadClass(className, extension);
+
+  }
```
ConfigurationTypeHelper calls this utility class now. It's certainly possible to move this method to a new class, I'm not sure which class is appropriate. Suggestions?

dlmarion commented on core/src/main/java/org/apache/accumulo/core/classloader/ContextClassLoaders.java:
```diff
@@ -0,0 +1,108 @@
+
+  public static final String CONTEXT_CLASS_LOADER_FACTORY = "general.context.class.loader.factory";
+
+  private static ContextClassLoaderFactory FACTORY;
```
I can synchronize the method with no penalty as it's only called once per server process.

dlmarion commented on core/src/main/java/org/apache/accumulo/core/classloader/ContextClassLoaders.java:
```diff
@@ -0,0 +1,108 @@
+   * @param conf
+   *          AccumuloConfiguration object
+   */
+  @SuppressWarnings("unchecked")
```
I removed the annotation as I can't move it closer.

dlmarion commented on core/src/main/java/org/apache/accumulo/core/classloader/ContextClassLoaders.java:
```diff
@@ -0,0 +1,108 @@
+        return;
+      }
+      try {
+        var factoryClass = Class.forName(factoryName);
```
Changed to `isEmpty()`

dlmarion commented on core/src/main/java/org/apache/accumulo/core/classloader/ContextClassLoaders.java:
```diff
@@ -0,0 +1,103 @@
+   * @throws RuntimeException
+   *           if contextName not configured
+   */
+  public static ClassLoader getClassLoader(String contextName) {
```
Made this change.

dlmarion commented on core/src/main/java/org/apache/accumulo/core/classloader/ContextClassLoaders.java:
```diff
@@ -0,0 +1,108 @@
+      return FACTORY.getClassLoader(contextName);
+    } catch (IllegalArgumentException e) {
+      LOG.error("ContextClassLoaderFactory is not configured for context: {}", contextName);
+      throw new RuntimeException(
+          "ContextClassLoaderFactory is not configured for context: " + contextName);
```
Made this change.

dlmarion commented on core/src/main/java/org/apache/accumulo/core/classloader/LegacyVFSContextClassLoaderFactory.java:
```diff
@@ -0,0 +1,93 @@
+          }, 60000, 60000);
+      LOG.debug("Context cleanup timer started at 60s intervals");
+    } catch (IOException e) {
+      throw new RuntimeException(e);
```
interesting, didn't know about that. Thanks.

dlmarion commented on core/src/test/java/org/apache/accumulo/core/classloader/URLClassLoaderFactory.java:
```diff
@@ -0,0 +1,55 @@
+
+public class URLClassLoaderFactory implements ContextClassLoaderFactory {
+
+  private static final String COMMA = ",";
```
I'm not sure that this is important as it's a test class

ctubbsii replied:
Oh right. I was thinking we'd provide this one as a reference imlementation built-in, for users. If it's just a test, user experience doesn't matter.

EdColeman replied:
Some times test cases are a good way to find code to build upon - especially if there is no reference implementation or examples in documentation.  Not saying that this instance needs to change, but more of a something to keep in mind.

dlmarion commented on core/src/test/java/org/apache/accumulo/core/classloader/URLClassLoaderFactory.java:
```diff
@@ -0,0 +1,55 @@
+      try {
+        urls.add(new URL(p));
+      } catch (MalformedURLException e) {
+        LOG.error("Error creating URL from classpath segment: " + p);
```
I removed the log

dlmarion commented on core/src/test/java/org/apache/accumulo/core/classloader/URLClassLoaderFactory.java:
```diff
@@ -0,0 +1,55 @@
+        urls.add(new URL(p));
+      } catch (MalformedURLException e) {
+        LOG.error("Error creating URL from classpath segment: " + p);
+        throw new RuntimeException("Error creating URL from classpath segment: " + p, e);
```
Changed to throw IAE

dlmarion commented:
@ctubbsii @keith-turner - I believe I have made most of the requested changes. I did not make the change regarding the InitParameters in the SPI class as there did not seem to be consensus there. I just re-tested everything to ensure that it's still working. Any last thoughts before I merge?
